jaxlib/xla/sharding.h

```diff
@@ -84,9 +84,11 @@ class NamedSharding : public Sharding {
     return logical_device_ids_;
   }
 
-  static nanobind::handle type() {
-    static auto type = nanobind::type<NamedSharding>();
-    return type;
+  static nanobind::handle & type() {
+    static absl::Mutex mu;
+    absl::MutexLock lock(&mu);
+    static nanobind::handle* type = new nanobind::handle(nanobind::type<NamedSharding>());
```

Comment:
Unfortunately the fix here isn't correct. This still has the same bug: we're holding the mutex that C++ uses to protect initialization of a static variable while calling into Python.

However, there's a simpler way. We can just compute these values eagerly instead of lazily, which is a faster thing to do anyway. See
https://github.com/jax-ml/jax/pull/27764

Reply:
Let me check your PR and run the repro test. I tried something like that before this PR's code but got still some other deadlock ...

Reply:
Actually, I'm not totally sure where there is a python call behind: nanobind::type<NamedSharding>(). It tries to fetch the type from the mapping check, IMO:
https://github.com/wjakob/nanobind/blob/9b3afa9dbdc23641daf26fadef7743e7127ff92f/include/nanobind/nb_types.h#L631-L633

https://github.com/wjakob/nanobind/blob/9b3afa9dbdc23641daf26fadef7743e7127ff92f/src/nb_type.cpp#L2055-L2061

https://github.com/wjakob/nanobind/blob/9b3afa9dbdc23641daf26fadef7743e7127ff92f/src/nb_type.cpp#L350-L385